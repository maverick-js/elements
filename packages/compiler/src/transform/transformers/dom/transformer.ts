import {
  $,
  createImports,
  isJsxRootNode,
  resetUniqueIdCount,
  splitImportsAndBody,
  transformTsNode,
} from '@maverick-js/ts';
import ts from 'typescript';

import {
  isMaverickCoreImportDeclaration,
  removeVirtualComponentImports,
} from '../../../parse/analysis';
import { Scope } from '../../../parse/ast';
import { parse } from '../../../parse/parse';
import { setupCustomElements } from '../shared/element';
import { Variables } from '../shared/variables';
import type { Transform, TransformData } from '../transformer';
import type { DomRuntime } from './runtime';
import { DomTransformState } from './state';
import { transform } from './transform';

export interface DomTransformOptions {
  /**
   * Whether to mark components with a static `element` property declaration to be
   * built as a custom element.
   */
  customElements?: boolean;
  /**
   * When set to `true`, the DOM runtime will connect the server-rendered HTML with the JS on
   * the client, making it interactive.
   *
   * The runtime does not require the component to be rendered on the server first but it's
   * recommended for the best performance.
   */
  hydratable?: boolean;
  /**
   * A single event listener is attached to the document to manage expensive events on multiple
   * child elements. Instead of adding individual listeners to each child, the event "bubbles" up
   * from the target element to the document, where the listener can capture and handle it. This
   * improves performance, especially in dynamic interfaces, by reducing the number of event
   * listeners.
   */
  delegateEvents?: boolean;
}

export function createDomTransform(options?: DomTransformOptions): Transform {
  return (data) => domTransform(data, options);
}

export function domTransform(
  { sourceFile }: TransformData,
  { customElements, hydratable, delegateEvents }: DomTransformOptions = {},
) {
  let state = new DomTransformState(null, { hydratable }),
    seenImports = false;

  function visit(this: ts.TransformationContext, node: ts.Node) {
    if (!seenImports && isMaverickCoreImportDeclaration(node)) {
      seenImports = true;
      return removeVirtualComponentImports(node, (name) => state.runtime.add(name));
    } else if (isJsxRootNode(node)) {
      const ast = parse(node),
        result = ast ? transform(ast, state.child(ast, new Scope())) : $.null;

      resetUniqueIdCount();

      return result;
    }

    return ts.visitEachChild(node, visit, this);
  }

  const transformedSourceFile = transformTsNode(sourceFile, (ctx) => visit.bind(ctx)),
    { imports, body } = splitImportsAndBody(transformedSourceFile);

  if (delegateEvents && state.delegatedEvents.size > 0) {
    body.push(createDelegateEventsStatement(state.delegatedEvents, state.runtime));
  }

  const statements: ts.Statement[] = [];

  const templates = createTemplateVariables(state);

  if (state.runtime.identifiers.length > 0) {
    statements.push(state.runtime.toImportDeclaration());
  }

  if (templates) statements.push(templates.toStatement());

  if (state.vars.module.length > 0) {
    statements.push(state.vars.module.toStatement());
  }

  const updatedSourceFile = $.updateSourceFile(transformedSourceFile, [
    ...imports,
    ...statements,
    ...state.renders,
    ...body,
  ]);

  return customElements
    ? setupCustomElements(updatedSourceFile, registerCustomElement, [createElementImport()])
    : updatedSourceFile;
}

const registerCustomElementExpression = $.call($.id('$$_define_custom_element'), [$.id('this')]);

function registerCustomElement() {
  return registerCustomElementExpression;
}

function createElementImport() {
  return createImports([$.id('$$_define_custom_element')], '@maverick-js/element');
}

function createTemplateVariables(state: DomTransformState) {
  const templates: [template: ts.Identifier, html: string, importNodes: boolean][] = [],
    vars = new Variables();

  state.walk(({ template, html, importNodes }) => {
    if (template && html) templates.push([template, html, importNodes]);
  });

  for (const [template, html, importNodes] of templates) {
    vars.create(template, state.runtime.createTemplate($.string(html), importNodes));
  }

  // Dedupe templates.
  for (let i = 0; i < templates.length; i++) {
    const templateA = templates[i];
    for (let j = i + 1; j < templates.length; j++) {
      const templateB = templates[j];
      if (templateA[1] === templateB[1] && templateA[2] === templateB[2]) {
        vars.update(templateB[0], templateA[0]);
      }
    }
  }

  return templates.length > 0 ? vars : null;
}

export function createDelegateEventsStatement(events: Set<string>, runtime: DomRuntime) {
  return $.createExpressionStatement(
    runtime.delegateEvents($.array(Array.from(events).map((type) => $.string(type)))),
  );
}
