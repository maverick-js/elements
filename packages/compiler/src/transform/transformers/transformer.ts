import type ts from 'typescript';

import type { AstNode } from '../../parse/ast';
import type { TransformContext } from '../transform';

export interface Transform {
  (data: TransformData): ts.SourceFile;
}

export interface TransformData {
  sourceFile: ts.SourceFile;
  nodes: AstNode[];
  ctx: TransformContext;
}

export interface StateTransform<State> {
  (node: AstNode, state: State): StateTransformResult;
}

export type StateTransformResult = ts.Expression | Array<ts.Expression | ts.Statement> | undefined;
