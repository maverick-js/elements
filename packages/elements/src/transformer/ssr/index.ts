import type { ASTSerializer } from '../transform';

export const ssr: ASTSerializer = {
  serialize(ast, ctx) {
    return '';
  },
};
