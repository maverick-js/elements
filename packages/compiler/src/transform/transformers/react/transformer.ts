import { isArray } from '@maverick-js/std';
import {
  $,
  addStatementsToBlock,
  type BlockUpdate,
  isJsxRootNode,
  replaceTsNode,
  resetUniqueIdCount,
  splitImportsAndBody,
  transformTsNode,
  updateBlock,
} from '@maverick-js/ts';
import ts from 'typescript';

import {
  isMaverickCoreImportDeclaration,
  removeVirtualComponentImports,
} from '../../../parse/analysis';
import { parse } from '../../../parse/parse';
import { createDelegateEventsStatement } from '../dom/transformer';
import { returnLastExpression } from '../shared/factory';
import type { Transform, TransformData } from '../transformer';
import { ReactTransformState } from './state';
import { transform } from './transform';

export interface ReactTransformOptions {
  /**
   * A single event listener is attached to the document to manage expensive events on multiple
   * child elements. Instead of adding individual listeners to each child, the event "bubbles" up
   * from the target element to the document, where the listener can capture and handle it. This
   * improves performance, especially in dynamic interfaces, by reducing the number of event
   * listeners.
   */
  delegateEvents?: boolean;
}

export function createReactTransform(options?: ReactTransformOptions): Transform {
  return (data) => reactTransform(data, options);
}

export function reactTransform(
  { sourceFile }: TransformData,
  { delegateEvents }: ReactTransformOptions = {},
) {
  let state = new ReactTransformState(null),
    seenImports = false,
    blocks: BlockUpdate[][] = [];

  function visit(this: ts.TransformationContext, node: ts.Node) {
    if (!seenImports && isMaverickCoreImportDeclaration(node)) {
      seenImports = true;
      return removeVirtualComponentImports(node, (name) => state.runtime.add(name));
    } else if (ts.isBlock(node) || ts.isModuleBlock(node)) {
      const updates: BlockUpdate[] = [];

      blocks.push(updates);
      const block = ts.visitEachChild(node, visit, this);
      blocks.pop();

      return updateBlock(block, updates);
    } else if (isJsxRootNode(node)) {
      const ast = parse(node),
        result = ast ? transform(ast, state.child(ast)) : $.null,
        currentBlock = blocks.at(-1);

      resetUniqueIdCount();

      if (currentBlock) {
        if (isArray(result)) {
          const statements = result.slice(0, -1);

          currentBlock.push({
            type: 'add',
            statements,
            before: node,
          });

          currentBlock.push({
            type: 'replace',
            old: node,
            new: result.at(-1)!,
          });
        } else {
          currentBlock.push({
            type: 'replace',
            old: node,
            new: result,
          });
        }

        // Return original block and transform above.
        return node;
      }

      if (isArray(result)) {
        returnLastExpression(result);
        return $.selfInvokedFn(result);
      }

      return result;
    }

    return ts.visitEachChild(node, visit, this);
  }

  const transformedSourceFile = transformTsNode(sourceFile, (ctx) => visit.bind(ctx)),
    { imports, body } = splitImportsAndBody(transformedSourceFile);

  if (delegateEvents && state.delegatedEvents.size > 0) {
    body.push(createDelegateEventsStatement(state.delegatedEvents, state.domRuntime));
  }

  const statements: ts.Statement[] = [];

  const runtimes = [state.runtime, state.domRuntime, state.ssrRuntime];
  for (const runtime of runtimes) {
    if (runtime.identifiers.length > 0) {
      statements.push(runtime.toImportDeclaration());
    }
  }

  if (state.module.vars.length > 0) {
    statements.push(state.module.vars.toStatement());
  }

  return $.updateSourceFile(sourceFile, [
    ...imports,
    ...statements,
    ...state.module.block,
    ...body,
  ]);
}
