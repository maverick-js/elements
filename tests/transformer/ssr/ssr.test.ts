import { transform } from 'maverick.js/transformer';

const t = (code: string) => transform(code, { generate: 'ssr' }).code;

it('should return sourcemap', () => {
  const result = transform(`<div></div>`, { filename: 'foo.tsx', sourcemap: true });
  expect(result.map).toBeDefined();
  expect(result.map!.sources[0]).toBe('foo.tsx');
});

it('should compile single JSX node', () => {
  const result = t(`<div></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile single self-closing JSX node', () => {
  const result = t(`<div />`);
  expect(result).toMatchSnapshot();
});

it('should compile multiple JSX nodes', () => {
  const result = t(`<div><span id="a"></span><span id="b"></span></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile fragment', () => {
  const result = t(`<><div id="a"></div><div id="b"></div></>`);
  expect(result).toMatchSnapshot();
});

it('should compile nested fragment', () => {
  const result = t(`<div id="root"><><div id="a"></div><div id="b"></div></></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile shorthand boolean attribute', () => {
  const result = t(`<button disabled />`);
  expect(result).toMatchSnapshot();
});

it('should compile static attributes', () => {
  const result = t(`<div class="foo bar" style="baz daz"></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile static attribute (number)', () => {
  const result = t(`<div foo={10}></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile static attribute (boolean)', () => {
  const result = t(`<div foo={true} bar={false}></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile static attribute (template string)', () => {
  const result = t(`<div foo={\`bar-baz\`} ></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile spread', () => {
  const result = t(`<div {...props} ></div>`);
  expect(result).toMatchInlineSnapshot(`
    "import { $$_spread, $$_ssr } from \\"maverick.js/ssr\\";

    const $$_templ = /* #__PURE__ */ [\\"<!$><div\\", \\"></div>\\"];
    $$_ssr($$_templ, $$_spread([props]))"
  `);
});

it('should compile spread with attributes', () => {
  const result = t(`<div a="0" b="1" c={id} {...props} d={true} e={10} {...propsTwo}></div>`);
  expect(result).toMatchSnapshot();
});

it('should merge classes between spreads', () => {
  const result = t(
    `<div $class:a={id} $class:b={false} class="foo bar" {...props} $class:c={0} {...propsTwo}></div>`,
  );
  expect(result).toMatchSnapshot();
});

it('should merge classes after spread', () => {
  const result = t(`<div {...props} $class:a={id} $class:b={false} class="foo bar"></div>`);
  expect(result).toMatchSnapshot();
});

it('should merge styles between spreads', () => {
  const result = t(
    `<div $style:a={id} $style:b={false} style="foo bar" {...props} $style:c={0} {...propsTwo}></div>`,
  );
  expect(result).toMatchSnapshot();
});

it('should merge styles after spread', () => {
  const result = t(
    `<div {...props} $style:a={id} $style:b={false} style="foo bar"  $style:c={0}></div>`,
  );
  expect(result).toMatchSnapshot();
});

it('should merge spreads', () => {
  const result = t(`<div {...props} {...propsTwo} {...propsThree} ></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile dynamic attribute', () => {
  const result = t(`<div foo={id}></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile observable attribute', () => {
  const result = t(`<div foo={id()}></div>`);
  expect(result).toMatchSnapshot();
});

it('should _not_ compile innerHTML expression', () => {
  const result = t(`<div $prop:innerHTML="baz"></div>`);
  expect(result).toMatchSnapshot();
});

it('should _not_ compile $prop expression', () => {
  const result = t(`<div $prop:fooBar="baz"></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile static class attribute', () => {
  const result = t(`<div class="foo bar"></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile dynamic class attribute', () => {
  const result = t(`<div class={id()}></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile $class expression', () => {
  const result = t(`<div $class:foo={true}></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile multiple $class expressions', () => {
  const result = t(
    `<div class="foo bar" $class:foo={true} $class:bar={false} $class:baz={true}></div>`,
  );
  expect(result).toMatchSnapshot();
});

it('should compile observable $class expression', () => {
  const result = t(`<div $class:foo={id()}></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile static style attribute', () => {
  const result = t(`<div style="foo: 1; bar: 2;"></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile dynamic style attribute', () => {
  const result = t(`<div style={id()}></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile $style expression', () => {
  const result = t(`<div $style:foo="bar"></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile dynamic $style expression', () => {
  const result = t(`<div $style:foo={id}></div>`);
  expect(result).toMatchSnapshot();
});

it('should group multiple static $style expressions', () => {
  const result = t(
    `<div style="foo: a;" $style:foo={"b"} $style:bar={true} $style:baz={id}></div>`,
  );
  expect(result).toMatchSnapshot();
});

it('should compile observable $style expression', () => {
  const result = t(`<div style={id()} $style:foo={id()}></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile $cssvar expression', () => {
  const result = t(`<div $cssvar:foo={10}></div>`);
  expect(result).toMatchSnapshot();
});

it('should group multiple static $cssvar expressions', () => {
  const result = t(
    `<div style="pre: 10" $cssvar:foo={10} $cssvar:bar={'align-content'} $cssvar:baz={id}></div>`,
  );
  expect(result).toMatchSnapshot();
});

it('should group multiple static $style and $cssvar expressions', () => {
  const result = t(
    `<div style="pre: 10" $style:baz={\`content\`} $style:boo={20} $cssvar:foo={10} $cssvar:bar={'align-content'}></div>`,
  );
  expect(result).toMatchSnapshot();
});

it('should compile dynamic $cssvar expression', () => {
  const result = t(`<div $cssvar:foo={id}></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile observable $cssvar expression', () => {
  const result = t(`<div $cssvar:foo-bar={id()}></div>`);
  expect(result).toMatchSnapshot();
});

it('should _not_ compile $use expression', () => {
  const result = t(`<div $use:directive={[1, 2, 3]}></div>`);
  expect(result).toMatchSnapshot();
});

it('should _not_ compile $ref expression', () => {
  const result = t(`<div $ref={(el) => {}}></div>`);
  expect(result).toMatchSnapshot();
});

it('should _not_ compile $on expression', () => {
  const result = t(`<div $on:foo={(e) => {}} />`);
  expect(result).toMatchSnapshot();
});

it('should _not_ compile $oncapture expression', () => {
  const result = t(`<div $oncapture:foo={(e) => {}} />`);
  expect(result).toMatchSnapshot();
});

it('should compile child expression', () => {
  const result = t(`<div>{id}</div>`);
  expect(result).toMatchSnapshot();
});

it('should compile observable child expression', () => {
  const result = t(`<div>{id() + 10}</div>`);
  expect(result).toMatchSnapshot();
});

it('should compile conditional element expression ', () => {
  const result = t(`<div id="a">{id > 10 && <div id="b"></div>}</div>`);
  expect(result).toMatchSnapshot();
});

it('should compile observable conditional element expression ', () => {
  const result = t(`<div id="a">{id() > 10 && <div id="b" $on:click={id()}></div>}</div>`);
  expect(result).toMatchSnapshot();
});

it('shoud merge duplicate templates', () => {
  const result = t(`
<div></div>
<div></div>
<div>Foo</div>
<div>Foo</div>
<div>Bar</div>
<span></span>
<span>Foo</span>
<span>Foo</span>
<span>Foo</span>
<span>Bar</span>
`);
  expect(result).toMatchSnapshot();
});
