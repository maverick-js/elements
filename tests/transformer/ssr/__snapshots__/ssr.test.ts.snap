// Vitest Snapshot v1

exports[`shoud merge duplicate templates 1`] = `
"import { $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<div></div>\\"],
  $$_templ_2 = /* #__PURE__ */ $$_templ,
  $$_templ_3 = /* #__PURE__ */ [\\"<div>Foo</div>\\"],
  $$_templ_4 = /* #__PURE__ */ $$_templ_3,
  $$_templ_5 = /* #__PURE__ */ [\\"<div>Bar</div>\\"],
  $$_templ_6 = /* #__PURE__ */ [\\"<span></span>\\"],
  $$_templ_7 = /* #__PURE__ */ [\\"<span>Foo</span>\\"],
  $$_templ_8 = /* #__PURE__ */ $$_templ_7,
  $$_templ_9 = /* #__PURE__ */ $$_templ_7,
  $$_templ_10 = /* #__PURE__ */ [\\"<span>Bar</span>\\"];

$$_ssr($$_templ)
$$_ssr($$_templ_2)
$$_ssr($$_templ_3)
$$_ssr($$_templ_4)
$$_ssr($$_templ_5)
$$_ssr($$_templ_6)
$$_ssr($$_templ_7)
$$_ssr($$_templ_8)
$$_ssr($$_templ_9)
$$_ssr($$_templ_10)
"
`;

exports[`should _not_ compile $on expression 1`] = `
"import { $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<!$><div></div>\\"];
$$_ssr($$_templ)"
`;

exports[`should _not_ compile $oncapture expression 1`] = `
"import { $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<!$><div></div>\\"];
$$_ssr($$_templ)"
`;

exports[`should _not_ compile $prop expression 1`] = `
"import { $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<!$><div></div>\\"];
$$_ssr($$_templ)"
`;

exports[`should _not_ compile $ref expression 1`] = `
"import { $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<!$><div></div>\\"];
$$_ssr($$_templ)"
`;

exports[`should _not_ compile $use expression 1`] = `
"import { $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<!$><div></div>\\"];
$$_ssr($$_templ)"
`;

exports[`should _not_ compile innerHTML expression 1`] = `
"import { $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<!$><div></div>\\"];
$$_ssr($$_templ)"
`;

exports[`should compile $class expression 1`] = `
"import { $$_classes, $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<!$><div\\", \\"></div>\\"];
$$_ssr($$_templ, $$_classes(\\"\\", { \\"foo\\": true }))"
`;

exports[`should compile $cssvar expression 1`] = `
"import { $$_styles, $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<div\\", \\"></div>\\"];
$$_ssr($$_templ, $$_styles(\\"\\", { \\"--foo\\": 10 }))"
`;

exports[`should compile $style expression 1`] = `
"import { $$_styles, $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<div\\", \\"></div>\\"];
$$_ssr($$_templ, $$_styles(\\"\\", { \\"foo\\": \\"bar\\" }))"
`;

exports[`should compile child expression 1`] = `
"import { $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<div><!$>\\", \\"</div>\\"];
$$_ssr($$_templ, id)"
`;

exports[`should compile conditional element expression  1`] = `
"import { $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<div id=\\\\\\"b\\\\\\"></div>\\"],
  $$_templ_2 = /* #__PURE__ */ [\\"<div id=\\\\\\"a\\\\\\"><!$>\\", \\"</div>\\"];
$$_ssr($$_templ_2, id > 10 && $$_ssr($$_templ))"
`;

exports[`should compile dynamic $cssvar expression 1`] = `
"import { $$_styles, $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<!$><div\\", \\"></div>\\"];
$$_ssr($$_templ, $$_styles(\\"\\", { \\"--foo\\": id }))"
`;

exports[`should compile dynamic $style expression 1`] = `
"import { $$_styles, $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<!$><div\\", \\"></div>\\"];
$$_ssr($$_templ, $$_styles(\\"\\", { \\"foo\\": id }))"
`;

exports[`should compile dynamic attribute 1`] = `
"import { $$_attr, $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<!$><div\\", \\"></div>\\"];
$$_ssr($$_templ, $$_attr(\\"foo\\", id))"
`;

exports[`should compile dynamic class attribute 1`] = `
"import { $$_classes, $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<!$><div\\", \\"></div>\\"];
$$_ssr($$_templ, $$_classes(id))"
`;

exports[`should compile dynamic style attribute 1`] = `
"import { $$_styles, $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<!$><div\\", \\"></div>\\"];
$$_ssr($$_templ, $$_styles(id))"
`;

exports[`should compile fragment 1`] = `
"import { $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<div id=\\\\\\"a\\\\\\"></div><div id=\\\\\\"b\\\\\\"></div>\\"];
$$_ssr($$_templ)"
`;

exports[`should compile multiple $class expressions 1`] = `
"import { $$_classes, $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<!$><div\\", \\"></div>\\"];
$$_ssr($$_templ, $$_classes(\\"foo bar\\", { \\"foo\\": true, \\"bar\\": false, \\"baz\\": true }))"
`;

exports[`should compile multiple JSX nodes 1`] = `
"import { $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<div><span id=\\\\\\"a\\\\\\"></span><span id=\\\\\\"b\\\\\\"></span></div>\\"];
$$_ssr($$_templ)"
`;

exports[`should compile nested fragment 1`] = `
"import { $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<div id=\\\\\\"root\\\\\\"><div id=\\\\\\"a\\\\\\"></div><div id=\\\\\\"b\\\\\\"></div></div>\\"];
$$_ssr($$_templ)"
`;

exports[`should compile observable $class expression 1`] = `
"import { $$_classes, $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<!$><div\\", \\"></div>\\"];
$$_ssr($$_templ, $$_classes(\\"\\", { \\"foo\\": id }))"
`;

exports[`should compile observable $cssvar expression 1`] = `
"import { $$_styles, $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<!$><div\\", \\"></div>\\"];
$$_ssr($$_templ, $$_styles(\\"\\", { \\"--foo-bar\\": id }))"
`;

exports[`should compile observable $style expression 1`] = `
"import { $$_styles, $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<!$><div\\", \\"></div>\\"];
$$_ssr($$_templ, $$_styles(id, { \\"foo\\": id }))"
`;

exports[`should compile observable attribute 1`] = `
"import { $$_attr, $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<!$><div\\", \\"></div>\\"];
$$_ssr($$_templ, $$_attr(\\"foo\\", id))"
`;

exports[`should compile observable child expression 1`] = `
"import { $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<div><!$>\\", \\"</div>\\"];
$$_ssr($$_templ, () => id() + 10)"
`;

exports[`should compile observable conditional element expression  1`] = `
"import { $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<!$><div id=\\\\\\"b\\\\\\"></div>\\"],
  $$_templ_2 = /* #__PURE__ */ [\\"<div id=\\\\\\"a\\\\\\"><!$>\\", \\"</div>\\"];
$$_ssr($$_templ_2, () => id() > 10 && $$_ssr($$_templ))"
`;

exports[`should compile shorthand boolean attribute 1`] = `
"import { $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<button disabled=\\\\\\"\\\\\\"></button>\\"];
$$_ssr($$_templ)"
`;

exports[`should compile single JSX node 1`] = `
"import { $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<div></div>\\"];
$$_ssr($$_templ)"
`;

exports[`should compile single self-closing JSX node 1`] = `
"import { $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<div></div>\\"];
$$_ssr($$_templ)"
`;

exports[`should compile spread with attributes 1`] = `
"import { $$_spread, $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<!$><div\\", \\"></div>\\"];
$$_ssr($$_templ, $$_spread([{ \\"a\\": \\"0\\", \\"b\\": \\"1\\", \\"c\\": id }, props, { \\"d\\": true, \\"e\\": 10 }, propsTwo]))"
`;

exports[`should compile static attribute (boolean) 1`] = `
"import { $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<div foo=\\\\\\"true\\\\\\" bar=\\\\\\"false\\\\\\"></div>\\"];
$$_ssr($$_templ)"
`;

exports[`should compile static attribute (number) 1`] = `
"import { $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<div foo=\\\\\\"10\\\\\\"></div>\\"];
$$_ssr($$_templ)"
`;

exports[`should compile static attribute (template string) 1`] = `
"import { $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<div foo=\\\\\\"bar-baz\\\\\\"></div>\\"];
$$_ssr($$_templ)"
`;

exports[`should compile static attributes 1`] = `
"import { $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<div class=\\\\\\"foo bar\\\\\\" style=\\\\\\"baz daz\\\\\\"></div>\\"];
$$_ssr($$_templ)"
`;

exports[`should compile static class attribute 1`] = `
"import { $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<div class=\\\\\\"foo bar\\\\\\"></div>\\"];
$$_ssr($$_templ)"
`;

exports[`should compile static style attribute 1`] = `
"import { $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<div style=\\\\\\"foo: 1; bar: 2\\\\\\"></div>\\"];
$$_ssr($$_templ)"
`;

exports[`should group multiple static $cssvar expressions 1`] = `
"import { $$_styles, $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<!$><div\\", \\"></div>\\"];
$$_ssr($$_templ, $$_styles(\\"pre: 10\\", { \\"--foo\\": 10, \\"--bar\\": \\"align-content\\", \\"--baz\\": id }))"
`;

exports[`should group multiple static $style and $cssvar expressions 1`] = `
"import { $$_styles, $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<div\\", \\"></div>\\"];
$$_ssr($$_templ, $$_styles(\\"pre: 10\\", { \\"baz\\": \`content\`, \\"boo\\": 20, \\"--foo\\": 10, \\"--bar\\": \\"align-content\\" }))"
`;

exports[`should group multiple static $style expressions 1`] = `
"import { $$_styles, $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<!$><div\\", \\"></div>\\"];
$$_ssr($$_templ, $$_styles(\\"foo: a;\\", { \\"foo\\": \\"b\\", \\"bar\\": true, \\"baz\\": id }))"
`;

exports[`should merge classes after spread 1`] = `
"import { $$_spread, $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<!$><div\\", \\"></div>\\"];
$$_ssr($$_templ, $$_spread([props, { \\"class\\": \\"foo bar\\", \\"$$class\\": { \\"a\\": id, \\"b\\": false } }]))"
`;

exports[`should merge classes between spreads 1`] = `
"import { $$_spread, $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<!$><div\\", \\"></div>\\"];
$$_ssr(
  $$_templ,
  $$_spread([{ \\"class\\": \\"foo bar\\", \\"$$class\\": { \\"a\\": id, \\"b\\": false } }, props, { \\"$$class\\": { \\"c\\": 0 } }, propsTwo]),
)"
`;

exports[`should merge spreads 1`] = `
"import { $$_spread, $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<!$><div\\", \\"></div>\\"];
$$_ssr($$_templ, $$_spread([props, propsTwo, propsThree]))"
`;

exports[`should merge styles after spread 1`] = `
"import { $$_spread, $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<!$><div\\", \\"></div>\\"];
$$_ssr($$_templ, $$_spread([props, { \\"style\\": \\"foo bar\\", \\"$$style\\": { \\"a\\": id, \\"b\\": false, \\"c\\": 0 } }]))"
`;

exports[`should merge styles between spreads 1`] = `
"import { $$_spread, $$_ssr } from \\"maverick.js/ssr\\";

const $$_templ = /* #__PURE__ */ [\\"<!$><div\\", \\"></div>\\"];
$$_ssr(
  $$_templ,
  $$_spread([{ \\"style\\": \\"foo bar\\", \\"$$style\\": { \\"a\\": id, \\"b\\": false } }, props, { \\"$$style\\": { \\"c\\": 0 } }, propsTwo]),
)"
`;
